### PR TITLE
Refactor `bind_value`

### PR DESCRIPTION
**Pull request by cyx-6**

### CODE REVIEW

File: python/tvm/script/builder/builder.py [hunk 1]
```diff
@@ -49,13 +49,13 @@ def get(self) -> Frame:
 DefType = TypeVar("DefType", bound=Object)
 
 
-def def_(name: str, var: DefType) -> DefType:
-    return _ffi_api.Def(name, var)  # pylint: disable=no-member # type: ignore
+def name_obj(name: str, var: DefType) -> DefType:
```
Comment (junrushao) [inline python/tvm/script/builder/builder.py]: just use `name` as a verb

```suggestion
def name(name: str, var: DefType) -> DefType:
```

File: python/tvm/script/builder/builder.py [hunk 2]
```diff
@@ -49,13 +49,13 @@ def get(self) -> Frame:
...
+    return _ffi_api.NameObj(name, var)  # pylint: disable=no-member # type: ignore
 
 
-def def_many(
+def name_obj_list(
```
Comment (junrushao) [inline python/tvm/script/builder/builder.py]:
```suggestion
def name_many(
```

File: python/tvm/script/builder/tir/var.py [hunk 3]
```diff
@@ -101,10 +101,11 @@ class Ptr_:
     def __getitem__(self, args):
         if not isinstance(args, tuple):
             args = (args,)
-        if len(args) == 1:
-            args = (args[0], "global")
-        return _ffi_api.Ptr(PrimType(args[0]().dtype), args[1])
+        ptr_type = args[0] if isinstance(args[0], str) else args[0]().dtype
```
Comment (junrushao) [inline python/tvm/script/builder/tir/var.py]:
```suggestion
        ptr_type = args[0]().dtype if callable(args[0]) else args[0]
```

File: python/tvm/script/builder/tir/var.py [hunk 4]
```diff
@@ -101,10 +101,11 @@ class Ptr_:
     def __getitem__(self, args):
         if not isinstance(args, tuple):
             args = (args,)
-        if len(args) == 1:
-            args = (args[0], "global")
-        return _ffi_api.Ptr(PrimType(args[0]().dtype), args[1])
```
Comment (junrushao) [inline python/tvm/script/builder/tir/var.py:106]: add an assertion:

```python
if len(args) not in [1, 2]:
  raise ValueError(...)
```

File: python/tvm/script/parse/var_table.py [hunk 5]
```diff
@@ -62,3 +62,9 @@ def add(self, var: str, value: Any):
 
     def get(self) -> Dict[str, Any]:
         return {key: values[-1] for key, values in self.name2value.items() if values}
+
+    def exist(self, value: Any):
+        for v in self.name2value.values():
+            if v is value:
```
Comment (junrushao) [inline python/tvm/script/parse/var_table.py:68]: i'm not 100% sure whether or not we should use `is` operator here but I don't have better idea for now. shall we deliberate a little bit?

File: python/tvm/script/parse/tir/tir.py [hunk 6]
```diff
@@ -22,31 +22,61 @@
...
+def bind_with_value(self: Parser, var_name: str, value: Any) -> Any:
+    if isinstance(value, (list, tuple)):
+        for i in range(len(value)):
+            bind_with_value(self, f"{var_name}_{i}", value[i])
+        return value
+    elif isinstance(value, (Buffer, Var)):
+        name(var_name, value)
+        return value
+    else:
+        self.report_error(f"Do not know how to bind type: {type(value)} in with statement")
```
Comment (junrushao) [inline python/tvm/script/parse/tir/tir.py]: Note: there is still a missing argument in `self.report_error`

Comment (cyx-6) [reply python/tvm/script/parse/tir/tir.py]: yes, but we do not have `node` in `bind_value`

File: python/tvm/script/parse/parser.py [hunk 7]
```diff
@@ -101,7 +101,7 @@ def eval_assign(
     ) -> Dict[str, Any]:
         var_values = eval_assign(self, target, source)
         for k, v in var_values.items():
-            var = bind_value(self, k, v)
+            var = bind_value(self, target, k, v)
```
Comment (junrushao) [inline python/tvm/script/parse/parser.py:104]: let's fix the type annotation for `bind_value` on line 100

Comment (cyx-6) [reply python/tvm/script/parse/parser.py:104]: updated

**Review by junrushao — APPROVED** [review]
LGTM

**Pull request merged**